[docs][proposal] Scaled prometheus pipeline first draft

Scott8440 opened the pull request:
## Summary

Design doc for a scaled prometheus pipeline using Thanos.

## Test Plan
N/a

Scott8440 committed:
[docs][proposal] Scaled prometheus pipeline first draft

Scott8440 committed:
[docs][proposal] Scaled prometheus pipeline first draft

xjtian committed:
Use different channel for artifact publish notifications (#2053)

Summary:
## Summary

- Avoid spamming the CI channel

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2053

Test Plan:
- Tested the payload with curl

## Additional Information

- [ ] This change is backwards-breaking

Reviewed By: mpgermano, tmdzk

Differential Revision: D22702585

Pulled By: xjtian

fbshipit-source-id: b958fb55bed974d97cccb34ec9b0296bb051d8bc

themarwhal committed:
Run 'go mod tidy' (#2052)

Summary:
<!--
    Tag your PR title with the components that it touches.
    E.g. "[lte][agw] Changeset" or "[orc8r][docker] ..."
-->

Committing result of running `go mod tidy`

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2052

Test Plan: CWAG unit tests

Reviewed By: mpgermano

Differential Revision: D22702814

Pulled By: themarwhal

fbshipit-source-id: 7ee224d1fb897a830ab383c465e4a79d80945d4f

karthiksubraveti committed:
Revert yarn lock

Summary: Revert yarn lock from Eric's change

Reviewed By: xjtian

Differential Revision: D22701141

fbshipit-source-id: 7928ca7b92754b7f7f117a3b177678ee309dd1e0

ekfuhrmann committed:
Began styling global dialogs using eNodeB detail edit/add dialog (#2010)

Summary:
# Overview
This PR focuses on applying global styles for the various dialogs we use. At the moment this PR is focusing solely on eNodeB detail dialogs, but the global styles are applied throughout magma. Future PR's will align other modals to the design.

## Dialog Modal
The global styles for the dialog box itself, including `Backdrop`, `DialogTitle`, `DialogContent`, and `DialogActions` have been added to our default styles and will propagate across the platform. In addition, I've added another option for users to close the modal via a close button in the top right corner.

Input fields have had their styles updated as well to better match that of the design. I've added a new `focus` and `hover` state, as well as removed some of the curve to the corners.

#### Before
![image](https://user-images.githubusercontent.com/8878152/88107079-84a0d880-cb74-11ea-8cfa-07f0ea32f8b5.png)

#### After
![image](https://user-images.githubusercontent.com/8878152/88107027-6fc44500-cb74-11ea-80b5-30a5d45142a2.png)

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2010

Reviewed By: xjtian

Differential Revision: D22669134

Pulled By: karthiksubraveti

fbshipit-source-id: fbe17c71fe1d5b7f50cabfac49971b7bc46a6450

jbryce committed:
Adds a GitHub action to check Signed-off-by (#2039)

Summary:
This includes a yaml file that calls a GitHub action each time a
PR is created or commits are updated. It pulls the commits from
the PR and ensures that each one has a Signed-off-by status.

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2039

Test Plan: This action has been running in magma/community and magma/magma-website. It initially adds a status check that reports on a PR

Reviewed By: tmdzk

Differential Revision: D22700703

Pulled By: xjtian

fbshipit-source-id: ead8c88e065601204b1d4b6bb89c94a2b61136aa

xjtian committed:
Short-circuit slack notify step (#2055)

Summary:
Summary
 ---

- Don't construct the slack message if we don't need to

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2055

Test Plan:
- yolo

## Additional Information

- [ ] This change is backwards-breaking

Reviewed By: tmdzk

Differential Revision: D22704106

Pulled By: xjtian

fbshipit-source-id: 1bbf775d4a450c1aca2bab65ee4e9b0befd9c371

ulaskozat committed:
temporarily removing tests from sanity (#2056)

Summary:
Signed-off-by: Ulas Kozat <kozat@fb.com>

Running tests after preamble test s1aptests/test_modify_mme_config_for_sanity.py would lead to getting stuck while running the sanity tests for LTE. The issue seems to be that the restart process for the AGW interferes with the first test scenario after the test s1aptests/test_modify_mme_config_for_sanity.py. Removing the preamble and postamble as well as the tests that require them and moving to a commented out section to revisit later.

As a side note: manually running the preamble and postamble and the test of interest works without any issues.

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2056

Test Plan: Run `make integ_test` in test_vm under `~/magma/lte/gateway/python/integ_tests` while running the traffic and dev VMs.

Reviewed By: ssanadhya

Differential Revision: D22704585

Pulled By: ulaskozat

fbshipit-source-id: 64a5c81781425674c61eb63d37b383868bd8eced

rckclmbr committed:
Bump fbcnms-ui version

Summary:
There's a fix I need for magma, but also testing to be sure publishing
the package works

Reviewed By: Scott8440

Differential Revision: D22705209

fbshipit-source-id: 81471e5358be1819cef86b89f61cd543da70f566

apbuteau commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+### Dealing with Rescaling
+
+When a user decides to rescale their deployment (either up or down), series will inevitably be mapped to different pushgateways. This will mean that series will be scraped by different prometheus servers as well. Note that this will not cause any issues when querying/alerting as Thanos can handle the same series existing across multiple servers. The only effect this will have is some temporary no ...
```
Scaling up is probably less of an issue because you could send brand new series to the new prometheus servers.  It's more on the scaling down where you would have to redistribute series to existing servers.  My guess is that you won't scale down though, so probably not an issue.

Scott8440 replied:
Consider the situation where you scale up because you're noticing performance degradation in the pushgateways (too much load). In this case you wouldn't just want to send new series to new gateways, but rather you'd want to redistribute all series evenly amongst the pushgateways. Additionally, since the hash should only be dependent on the `gatewayID` and the number of pushgateways, we wouldn't be able to send "new" series to the newly scaled up pushgateways.

apbuteau commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+### Pushgateway Redundancy
+
+We can leverage the addition of multiple gateways to provide failure protection. As discussed above, it's not a significant problem if some series get shifted to a different prometheus server. Because of this, we can protect metrics against a pushgateway failure by simply pushing metrics to the next gateway in the hash circle if the first request fails. While I do not expect this to happen much i ...
```
A mitigation for this would be to use a queue service (e.g. Amazon SQS, Apache Kafka) to provide some fault tolerance if pushgateways go down.

Scott8440 replied:
Could definitely be something worth looking into. Would need to weigh the added complexity and costs of course. At first glance I feel like it probably wouldn't be worth it since the problem it solves (small number of datapoints in a different server) would be minimal at most but I'll keep it in mind.

pshelar committed:
mobilityd: dhcp_client: implement dhcp lease (#2051)

Summary:
Following patch add support for DHCP lease to support dynamic
ip address allocation via DHCP. DHCP state monitoring is done
in separate monitoring thread in mobilityD.

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2051

Test Plan: `make test` and `make integ_test`

Reviewed By: ardzoht

Differential Revision: D22704025

Pulled By: pshelar

fbshipit-source-id: bdbb72bf0a0af0d3bfa53f9638f669e2b48bc963

ssanadhya committed:
change Apache-license to BSD-like license (#2050)

Summary:
On behalf of OAI team, switching Apache-license to BSD-like license for
- lte/gateway/c/oai/
- ci-scripts/
- lte/gateway/deploy/roles/magma_oai/
- lte/protos/oai/spgw_state.proto

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2050

Test Plan:
`fab integ_test`

## Commit from Converged-MME repo
95734efb0       Raphael Defosseux       Fri Jun 26 10:34:01 2020 +0200  [OAI]: change Apache-license to BSD-like license

Reviewed By: ulaskozat, themarwhal, mpgermano

Differential Revision: D22703189

Pulled By: ssanadhya

fbshipit-source-id: 512300731695e6029fbcea06171b9e5f5f10a6d9

xjtian committed:
Fix fossa CI task (#2059)

Summary:
- Use buildpack container and add sudo to some commands

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2059

Test Plan:
https://app.circleci.com/pipelines/github/facebookincubator/magma/7181/workflows/e5eb9f0a-2a7e-414c-a1db-cae9b41561b8/jobs/56588

## Additional Information

- [ ] This change is backwards-breaking

Reviewed By: koolzz

Differential Revision: D22707742

Pulled By: xjtian

fbshipit-source-id: c42f0270e07bcba8252b5b6a9e6e6b1d48eae032

emakeev committed:
Switch EAP service to glog (#2058)

Summary:
Switch EAP service to glog

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2058

Test Plan:
make precommit on FeG

## Additional Information

Reviewed By: uri200

Differential Revision: D22712654

Pulled By: emakeev

fbshipit-source-id: 18b01b30f14cef842529ef9ecd60454826e4bbcd

xjtian commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+Thanos provides a solution to this problem in the [Ruler](https://thanos.io/components/rule.md/) component. It accepts rules files in the same format as prometheus, and evaluates the rules using a Thanos Querier which has access to all servers. Feature parity not documented (POST request to reload configuration files) validated by investigating [source code](https://github.com/thanos-io/thanos/bl ...
+
+To keep our alerting system all we have to do is configure the Thanos Ruler to use the alerting rules files managed by prometheus-configmanager, and remove the rules files from the individual prometheus servers.
```
Leave a pointer to the helm chart where this can be customized, or how we'd do this if the integration point doesn't exist already. Just have to make sure the Ruler component can accept an existing PVC to mount.

xjtian commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+![newMetricsPipeline.png](./newMetricsPipeline.png)
+
+To avoid making performance worse, we of course need to ensure that for each series, it is always sent to the same pushgateway and prometheus server. To do that, we need to implement some form of hashing. Preferably, this would all be done on the controllers, to avoid having to introduce a new “distributor” component. 
```
It seems to me there's a fundamental mismatch between the approach you're proposing and Thanos's philosphy for scaling timeseries queries and storage. From https://thanos.io/design.md/#scaling:

> Overall, first-class horizontal sharding is possible but will not be considered for the time being since there’s no evidence that it is required in practical setups.

I think you're trying to wrap a horizontal sharding solution around Thanos on the query side while trying to sidestep the things that make something like this difficult (e.g. consistent hashing, resharding) and this approach is introducing a lot of potentially fragile coupling further downstream in your proposed design.

I think we should try to run with the native scaling model for Thanos until we hit a wall with that (which may actually never happen). If we leverage object storage, we get to move the compute load of querying from the Prom server to the stateless Querier component which can scale horizontally far easier than Prometheus. Let's make this our story for the metrics stack:

- If you want to run on-prem on bare metal, don't deploy Thanos and live with your single Prometheus server.
- If you can deploy Openstack onto your BM, there's a compatible object storage solution.
- If you deploy into a public cloud, use S3 etc, which at massive data scale should actually save on the TCO side compared to EFS.

ssanadhya committed:
Format c/oai directory with clang-format (#2057)

Summary:
Applied the clang-format style file in root directory to all C/C++ files under c/oai.

Also fixed the log level in mme.yml to INFO

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2057

Test Plan: fab integ_test from magma/lte/gateway/

Reviewed By: ulaskozat

Differential Revision: D22709210

Pulled By: ssanadhya

fbshipit-source-id: 57621093dbbff15d6470d49aaf73bf1e2e573225

xjtian commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+### Dealing with Rescaling
+
+When a user decides to rescale their deployment (either up or down), series will inevitably be mapped to different pushgateways. This will mean that series will be scraped by different prometheus servers as well. Note that this will not cause any issues when querying/alerting as Thanos can handle the same series existing across multiple servers. The only effect this will have is some temporary no ...
```
> Thanos can handle the same series existing across multiple servers Thanos can handle the same series existing across multiple servers

This is actually a bit of an open question for me; Thanos can deduplicate datapoints from Prometheus servers that are deploiyed as HA replicas (same scrape configs), but all the architecture diagrams of federated querying I've seen have remote Prometheus servers that are separate and disjoint. Maybe the diagrams are simplified.

But anyways, I think we should roll with object storage and not worry about this.

xjtian commented on docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md:
```diff
@@ -0,0 +1,139 @@
+
+### Pushgateway Redundancy
+
+We can leverage the addition of multiple gateways to provide failure protection. As discussed above, it's not a significant problem if some series get shifted to a different prometheus server. Because of this, we can protect metrics against a pushgateway failure by simply pushing metrics to the next gateway in the hash circle if the first request fails. While I do not expect this to happen much i ...
```
Workloads in k8s can be interrupted for a variety of unpredictable reasons - node goes down, node is scheduled for an upgrade, pre-emption, etc. So we should build the assumption of ephemerality into the design of the system.

As I understand it, Prometheus's tsdb is append-only so it must read the datapoints for a single timeseries in order. With this constraint, I don't see any reasonable way around building some sort of consistent hashring for horizontally scaled pushgateways. You should put the hashring behind a load balancer to isolate the implementation complexity from the rest of the system; in other words, this is functionality that should live in pushgateway, not metricsd.

If you want to do some experiments to see how far we can get with a single pushgateway, let's anchor on an upper bound of 10 million datapoints per minute for this half. That's `100k devices * 100 datapoints` per minute. If it looks like we can survive on a single pushgateway at this scale, you can push the hashring out to H1 if integrating Thanos will take you most of the rest of this half.

xjtian reviewed (changes requested):
On the query-side, let's leverage Thanos out of the box to start. Lots of folks are running it at mind-boggling scale without a horizontally sharded Prometheus cluster, so I can't imagine we'd hit limits with this approach anytime soon.

xjtian commented:
PS I found https://www.slideshare.net/ThomasRiley45/scaling-prometheus-on-kubernetes-with-thanos a pretty useful slide deck

xjtian committed:
First stab at CODEOWNERS for the repo (#2060)

Summary:
Signed-off-by: Jacky Tian <xjtian@fb.com>

- Took a few creative liberties with the spreadsheet

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2060

Test Plan:
yolo

## Additional Information

- [ ] This change is backwards-breaking

Reviewed By: tmdzk

Differential Revision: D22721688

Pulled By: xjtian

fbshipit-source-id: 311a00872fa7e0ffbf96a9bac3cb0f46d58b8f5e

aclave1 committed:
BE(fbcnms): Remove useForm from fbcnms/hooks

Summary: It's unused by any other team and it's causing the UI to pull in unneeded dependencies.

Reviewed By: rckclmbr, cjminer505

Differential Revision: D22608238

fbshipit-source-id: 2782997e8d2815d60b83b0ab0426c60c2502bc76

ardzoht committed:
SyncRPC adding close_connection check (#2064)

Summary:
- Adding try/except check for close_connection as SyncRPC complains when connection is already closed

Pull Request resolved: https://github.com/facebookincubator/magma/pull/2064

Test Plan:
- tested on feg02 on mcp2 setup
- running make test

## Additional Information

- [ ] This change is backwards-breaking

Reviewed By: koolzz

Differential Revision: D22737940

Pulled By: ardzoht

fbshipit-source-id: 3b5645f08228465ef0bed8560e014fe349e4f898

ssanadhya committed:
[Copyright] Update copyright (#1994)

119Vik commented:
Can one of the admins verify this patch?

karthiksubraveti committed:
Plug EventTable into top level components (#2028)

Signed-off-by: Karthik Subraveti <karthikshyam@gmail.com>

xjtian committed:
[codeowners] Add xjtian to nms codeowners

Signed-off-by: Jacky Tian <xjtian@fb.com>

xjtian committed:
[ci] Re-enable docusaurus job (#2070)

Signed-off-by: Jacky Tian <xjtian@fb.com>

ssanadhya committed:
Remove extra comment in lte traffic server (#2072)

Signed-off-by: Shruti Sanadhya <ssanadhya@fb.com>

themarwhal committed:
Clean up SessionD's proto interfaces with RAT specific messages (#2063)

Signed-off-by: Marie Bremner <marwhal@fb.com>

HannaFar committed:
Add Gateway create and edit (#2009)

Signed-off-by: HannaFar <hannafarag159@gmail.com>

karthiksubraveti committed:
Remove package dependenciesSigned-off-by: Karthik Subraveti <karthikshyam@gmail.com> (#2074)

Signed-off-by: Karthik Subraveti <karthikshyam@gmail.com>

koolzz committed:
Update the CWF check quota return JSON (#2080)

Signed-off-by: Nick Yurchenko <koolzz@fb.com>

pshelar committed:
[AGW] pipelineD: add yml config for non-nat uplink port (#2079)

This configuration would be only used in Non-Modes.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

tmdzk committed:
[DOC]Remove all refs to facebookincubator.github.io (#2075)

Signed-off-by: Timothée Dzik <timdzik@fb.com>

jaredmullane committed:
Update Product Specifications Document  (#2084)

Signed-off-by: Jared Mullane <jmullane@fb.com>

shaddi committed:
[lte] Remove magma_oai vm (#2085)

Signed-off-by: Shaddi Hasan <shasan@fb.com>

tmdzk committed:
[CI]Implement xwfm test (#2069)

Signed-off-by: Timothée Dzik <timdzik@fb.com>

Commit:
Fix service303 registry error (#2086)

Signed-off-by: mgermano <mgermano@fb.com>

xjtian committed:
[nghttpx] Bring back nghttpx Dockerfile for now (#2092)

Signed-off-by: Jacky Tian <xjtian@fb.com>

119Vik committed:
fixed removeLastUITestCases.sed file (#2094)

Signed-off-by: Vitalii <vitalij.kostenko@gmail.com>

Scott8440 committed:
v2 proposal. Address comments and remove horizontal scaling in favor of object storage based optimizations

Signed-off-by: Scott8440 <scott8440@gmail.com>

Scott8440 committed:
merge

Signed-off-by: Scott8440 <scott8440@gmail.com>

Scott8440 commented:
please ignore while I fix this :)
